cirq/study/sweeps.py

```diff
@@ -101,6 +101,28 @@ def __iter__(self) -> Iterator[resolver.ParamResolver]:
         for params in self.param_tuples():
             yield resolver.ParamResolver(collections.OrderedDict(params))
 
+    def __getitem__(self, val) -> Union[Params, Iterator[Params]]:
```

Comment:
I think if we slice we should get back a `Sweep`, not just an `Iterator[Params]`. This is more consistent with other types, for example slicing a list gives a list. Also, getting a single index should return `resolver.ParamResolver` since if you just iterate over the sweep you get `Iterator[resolver.ParamResolver]` (see `__init__` above). For the latter, I think you just need to pass `self` instead of `self.param_tuples()` as the first arg to `itertools.islice`.

Also, note that you can use `an @overload` decorator to make the types more precise here so that mypy knows getting a single index returns one type, while getting a slice returns a Sweep (for reference, see how we do this in [`Circuit`](https://github.com/quantumlib/Cirq/blob/3bb71d11042543e26d3e6e4d0dc784f03eef4a5d/cirq/circuits/circuit.py#L177)).

Reply:
I'm returning a sweep now, but I'm not exactly sure about this @overload, can you give me a more concrete example ?